File: src/lib/scheduler.ts
export interface ProcessInput {
  id: string;
  arrivalTime: number;
  burstTime: number;
  priority: number;
}

export interface ExecBlock {
  processId: string;
  startTime: number;
  endTime: number;
}

export interface ProcessMetrics {
  id: string;
  completionTime: number;
  turnaroundTime: number;
  waitingTime: number;
  responseTime: number;
}

export interface SimulationResult {
  executionLog: ExecBlock[];
  metrics: ProcessMetrics[];
  avgTurnaroundTime: number;
  avgWaitingTime: number;
  avgResponseTime: number;
  cpuUtilization: number;
  throughput: number;
  contextSwitches: number;
}

export type Algorithm =
  | "FCFS"
  | "SJF"
  | "SRTF"
  | "Priority"
  | "RR"
  | "MLFQ";

const computeMetrics = (
  processes: ProcessInput[],
  executionLog: ExecBlock[]
): SimulationResult => {
  const completionTimes: Record<string, number> = {};
  const firstExecTimes: Record<string, number> = {};

  for (const block of executionLog) {
    if (block.processId !== "IDLE") {
      completionTimes[block.processId] = Math.max(
        completionTimes[block.processId] || 0,
        block.endTime
      );
      if (!(block.processId in firstExecTimes)) {
        firstExecTimes[block.processId] = block.startTime;
      }
    }
  }

  const metrics: ProcessMetrics[] = processes.map((p) => {
    const ct = completionTimes[p.id] || 0;
    const tat = ct - p.arrivalTime;
    const wt = tat - p.burstTime;
    const rt = (firstExecTimes[p.id] ?? p.arrivalTime) - p.arrivalTime;
    return {
      id: p.id,
      completionTime: ct,
      turnaroundTime: Math.max(0, tat),
      waitingTime: Math.max(0, wt),
      responseTime: Math.max(0, rt),
    };
  });

  const n = metrics.length || 1;
  const avgTurnaroundTime = metrics.reduce((a, m) => a + m.turnaroundTime, 0) / n;
  const avgWaitingTime = metrics.reduce((a, m) => a + m.waitingTime, 0) / n;
  const avgResponseTime = metrics.reduce((a, m) => a + m.responseTime, 0) / n;

  // Compress consecutive blocks of the same process
  const compressedLog: ExecBlock[] = [];
  for (const block of executionLog) {
    if (
      compressedLog.length > 0 &&
      compressedLog[compressedLog.length - 1].processId === block.processId &&
      compressedLog[compressedLog.length - 1].endTime === block.startTime
    ) {
      compressedLog[compressedLog.length - 1].endTime = block.endTime;
    } else {
      if (block.endTime > block.startTime) {
        compressedLog.push({ ...block });
      }
    }
  }

  // CPU Utilization
  const totalTime = compressedLog.length > 0 ? compressedLog[compressedLog.length - 1].endTime : 0;
  const idleTime = compressedLog
    .filter((b) => b.processId === "IDLE")
    .reduce((a, b) => a + (b.endTime - b.startTime), 0);
  const cpuUtilization = totalTime > 0 ? ((totalTime - idleTime) / totalTime) * 100 : 0;

  // Throughput
  const throughput = totalTime > 0 ? processes.length / totalTime : 0;

  // Context switches (count process transitions, ignoring IDLE)
  let contextSwitches = 0;
  let lastProcess: string | null = null;
  for (const block of compressedLog) {
    if (block.processId !== "IDLE") {
      if (lastProcess !== null && lastProcess !== block.processId) {
        contextSwitches++;
      }
      lastProcess = block.processId;
    }
  }

  return {
    executionLog: compressedLog,
    metrics,
    avgTurnaroundTime,
    avgWaitingTime,
    avgResponseTime,
    cpuUtilization,
    throughput,
    contextSwitches,
  };
};

export const simulateFCFS = (processes: ProcessInput[]): SimulationResult => {
  let currentTime = 0;
  const executionLog: ExecBlock[] = [];

  const sorted = [...processes].sort(
    (a, b) =>
      a.arrivalTime - b.arrivalTime ||
      parseInt(String(a.id).replace(/\D/g, "") || "0") -
        parseInt(String(b.id).replace(/\D/g, "") || "0")
  );

  for (const p of sorted) {
    if (currentTime < p.arrivalTime) {
      executionLog.push({ processId: "IDLE", startTime: currentTime, endTime: p.arrivalTime });
      currentTime = p.arrivalTime;
    }
    executionLog.push({ processId: p.id, startTime: currentTime, endTime: currentTime + p.burstTime });
    currentTime += p.burstTime;
  }

  return computeMetrics(processes, executionLog);
};

export const simulateSJF = (processes: ProcessInput[]): SimulationResult => {
  let currentTime = 0;
  const executionLog: ExecBlock[] = [];
  const remaining = [...processes].map((p) => ({ ...p }));

  while (remaining.length > 0) {
    const arrived = remaining.filter((p) => p.arrivalTime <= currentTime);

    if (arrived.length === 0) {
      const nextArrival = Math.min(...remaining.map((p) => p.arrivalTime));
      executionLog.push({ processId: "IDLE", startTime: currentTime, endTime: nextArrival });
      currentTime = nextArrival;
      continue;
    }

    arrived.sort((a, b) => a.burstTime - b.burstTime || a.arrivalTime - b.arrivalTime);
    const p = arrived[0];

    executionLog.push({ processId: p.id, startTime: currentTime, endTime: currentTime + p.burstTime });
    currentTime += p.burstTime;

    remaining.splice(
      remaining.findIndex((r) => r.id === p.id),
      1
    );
  }

  return computeMetrics(processes, executionLog);
};

export const simulatePriority = (processes: ProcessInput[]): SimulationResult => {
  let currentTime = 0;
  const executionLog: ExecBlock[] = [];
  const remaining = [...processes].map((p) => ({ ...p }));

  while (remaining.length > 0) {
    const arrived = remaining.filter((p) => p.arrivalTime <= currentTime);

    if (arrived.length === 0) {
      const nextArrival = Math.min(...remaining.map((p) => p.arrivalTime));
      executionLog.push({ processId: "IDLE", startTime: currentTime, endTime: nextArrival });
      currentTime = nextArrival;
      continue;
    }

    arrived.sort((a, b) => a.priority - b.priority || a.arrivalTime - b.arrivalTime);
    const p = arrived[0];

    executionLog.push({ processId: p.id, startTime: currentTime, endTime: currentTime + p.burstTime });
    currentTime += p.burstTime;

    remaining.splice(
      remaining.findIndex((r) => r.id === p.id),
      1
    );
  }

  return computeMetrics(processes, executionLog);
};

export const simulateSRTF = (processes: ProcessInput[]): SimulationResult => {
  let currentTime = 0;
  const executionLog: ExecBlock[] = [];
  const remaining = processes.map((p) => ({ ...p, remainingTime: p.burstTime }));

  while (remaining.length > 0) {
    const arrived = remaining.filter((p) => p.arrivalTime <= currentTime);

    if (arrived.length === 0) {
      const nextArrival = Math.min(...remaining.map((p) => p.arrivalTime));
      executionLog.push({ processId: "IDLE", startTime: currentTime, endTime: nextArrival });
      currentTime = nextArrival;
      continue;
    }

    arrived.sort((a, b) => a.remainingTime - b.remainingTime || a.arrivalTime - b.arrivalTime);
    const p = arrived[0];

    executionLog.push({ processId: p.id, startTime: currentTime, endTime: currentTime + 1 });
    p.remainingTime -= 1;
    currentTime += 1;

    if (p.remainingTime === 0) {
      remaining.splice(
        remaining.findIndex((r) => r.id === p.id),
        1
      );
    }
  }

  return computeMetrics(processes, executionLog);
};

export const simulateRR = (processes: ProcessInput[], timeQuantum: number): SimulationResult => {
  let currentTime = 0;
  const executionLog: ExecBlock[] = [];
  const remaining = processes
    .map((p) => ({ ...p, remainingTime: p.burstTime }))
    .sort((a, b) => a.arrivalTime - b.arrivalTime);

  const queue: typeof remaining = [];
  let remainingProcesses = [...remaining];

  const enqueueArrived = (time: number) => {
    const arrived = remainingProcesses.filter((p) => p.arrivalTime <= time);
    arrived.sort((a, b) => a.arrivalTime - b.arrivalTime);
    for (const p of arrived) {
      queue.push(p);
      remainingProcesses = remainingProcesses.filter((rp) => rp.id !== p.id);
    }
  };

  enqueueArrived(currentTime);

  while (queue.length > 0 || remainingProcesses.length > 0) {
    if (queue.length === 0) {
      const nextArrival = remainingProcesses[0].arrivalTime;
      executionLog.push({ processId: "IDLE", startTime: currentTime, endTime: nextArrival });
      currentTime = nextArrival;
      enqueueArrived(currentTime);
      continue;
    }

    const p = queue.shift()!;
    const execTime = Math.min(p.remainingTime, timeQuantum);

    executionLog.push({ processId: p.id, startTime: currentTime, endTime: currentTime + execTime });
    currentTime += execTime;
    p.remainingTime -= execTime;

    enqueueArrived(currentTime);

    if (p.remainingTime > 0) {
      queue.push(p);
    }
  }

  return computeMetrics(processes, executionLog);
};

export const simulateMLFQ = (
  processes: ProcessInput[],
  q0Quantum: number = 2,
  q1Quantum: number = 4
): SimulationResult => {
  let currentTime = 0;
  const executionLog: ExecBlock[] = [];

  const remainingProcesses = processes
    .map((p) => ({ ...p, remainingTime: p.burstTime }))
    .sort((a, b) => a.arrivalTime - b.arrivalTime);

  const q0: typeof remainingProcesses = [];
  const q1: typeof remainingProcesses = [];
  const q2: typeof remainingProcesses = [];
  let notYetArrived = [...remainingProcesses];

  const enqueueArrived = (time: number) => {
    const arrived = notYetArrived.filter((p) => p.arrivalTime <= time);
    arrived.sort((a, b) => a.arrivalTime - b.arrivalTime);
    for (const p of arrived) {
      q0.push(p);
      notYetArrived = notYetArrived.filter((rp) => rp.id !== p.id);
    }
  };

  enqueueArrived(currentTime);

  while (q0.length > 0 || q1.length > 0 || q2.length > 0 || notYetArrived.length > 0) {
    if (q0.length === 0 && q1.length === 0 && q2.length === 0) {
      const nextArrival = Math.min(...notYetArrived.map((p) => p.arrivalTime));
      executionLog.push({ processId: "IDLE", startTime: currentTime, endTime: nextArrival });
      currentTime = nextArrival;
      enqueueArrived(currentTime);
      continue;
    }

    if (q0.length > 0) {
      const p = q0.shift()!;
      const execTime = Math.min(p.remainingTime, q0Quantum);

      executionLog.push({ processId: p.id, startTime: currentTime, endTime: currentTime + execTime });
      currentTime += execTime;
      p.remainingTime -= execTime;

      enqueueArrived(currentTime);

      if (p.remainingTime > 0) {
        q1.push(p);
      }
    } else if (q1.length > 0) {
      const p = q1.shift()!;
      let t = 0;
      while (t < q1Quantum && p.remainingTime > 0) {
        executionLog.push({ processId: p.id, startTime: currentTime, endTime: currentTime + 1 });
        currentTime += 1;
        p.remainingTime -= 1;
        t += 1;

        enqueueArrived(currentTime);
        if (q0.length > 0) break;
      }
      if (p.remainingTime > 0) {
        if (t < q1Quantum) {
          q1.unshift(p); // Preempted by Q0 before finishing quantum, resume in Q1
        } else {
          q2.push(p); // Finished quantum, demote to Q2
        }
      }
    } else if (q2.length > 0) {
      const p = q2.shift()!;
      executionLog.push({ processId: p.id, startTime: currentTime, endTime: currentTime + 1 });
      currentTime += 1;
      p.remainingTime -= 1;

      enqueueArrived(currentTime);
      if (p.remainingTime > 0) {
        q2.unshift(p); // Preempted or just continuing FCFS tick-by-tick
      }
    }
  }

  return computeMetrics(processes, executionLog);
};

export const runSimulation = (
  processes: ProcessInput[],
  algo: Algorithm,
  quantumMap: { rr: number; q0: number; q1: number }
): SimulationResult => {
  if (processes.length === 0) {
    return {
      executionLog: [],
      metrics: [],
      avgTurnaroundTime: 0,
      avgWaitingTime: 0,
      avgResponseTime: 0,
      cpuUtilization: 0,
      throughput: 0,
      contextSwitches: 0,
    };
  }

  switch (algo) {
    case "FCFS":
      return simulateFCFS(processes);
    case "SJF":
      return simulateSJF(processes);
    case "SRTF":
      return simulateSRTF(processes);
    case "Priority":
      return simulatePriority(processes);
    case "RR":
      return simulateRR(processes, quantumMap.rr);
    case "MLFQ":
      return simulateMLFQ(processes, quantumMap.q0, quantumMap.q1);
    default:
      return simulateFCFS(processes);
  }
};

// Helper: get process states at a given tick for visualization
export function getStateAtTick(
  tick: number,
  processes: ProcessInput[],
  executionLog: ExecBlock[]
): {
  running: string | null;
  readyQueue: string[];
  completed: string[];
  notArrived: string[];
} {
  const currentBlock = executionLog.find((b) => b.startTime <= tick && tick < b.endTime);
  const running =
    currentBlock && currentBlock.processId !== "IDLE" ? currentBlock.processId : null;

  const executedTime: Record<string, number> = {};
  for (const block of executionLog) {
    if (block.processId === "IDLE") continue;
    const effectiveEnd = Math.min(block.endTime, tick);
    if (effectiveEnd > block.startTime) {
      executedTime[block.processId] =
        (executedTime[block.processId] || 0) + (effectiveEnd - block.startTime);
    }
  }

  const completed: string[] = [];
  const readyQueue: string[] = [];
  const notArrived: string[] = [];

  for (const p of processes) {
    const executed = executedTime[p.id] || 0;
    if (executed >= p.burstTime) {
      completed.push(p.id);
    } else if (p.arrivalTime > tick) {
      notArrived.push(p.id);
    } else if (p.id !== running) {
      readyQueue.push(p.id);
    }
  }

  return { running, readyQueue, completed, notArrived };
}

File: scheduler.py
"""
CPU Scheduling Algorithms - FCFS, SJF, SRTF, Priority, RR, MLFQ
With metrics: Response Time, CPU Utilization, Throughput, Context Switches
"""
from dataclasses import dataclass, field
from typing import List, Dict, Optional
from copy import deepcopy


@dataclass
class ProcessInput:
    id: str
    arrival_time: int
    burst_time: int
    priority: int = 1


@dataclass
class ExecBlock:
    process_id: str
    start_time: int
    end_time: int


@dataclass
class ProcessMetrics:
    id: str
    completion_time: int
    turnaround_time: int
    waiting_time: int
    response_time: int


@dataclass
class SimulationResult:
    execution_log: List[ExecBlock]
    metrics: List[ProcessMetrics]
    avg_turnaround_time: float
    avg_waiting_time: float
    avg_response_time: float
    cpu_utilization: float
    throughput: float
    context_switches: int


ALGORITHMS = ["FCFS", "SJF", "SRTF", "Priority", "RR", "MLFQ"]

ALGORITHM_INFO = {
    "FCFS": {
        "name": "First Come First Served",
        "type": "Non-Preemptive",
        "desc": "Processes execute in arrival order. Simple but can cause the convoy effect where short processes wait behind long ones.",
    },
    "SJF": {
        "name": "Shortest Job First",
        "type": "Non-Preemptive",
        "desc": "Selects the process with the smallest burst time. Optimal for minimizing average waiting time but requires advance knowledge of burst times.",
    },
    "SRTF": {
        "name": "Shortest Remaining Time First",
        "type": "Preemptive",
        "desc": "Preemptive SJF — at each tick, the process with the shortest remaining burst runs. Lowest average wait time but high context-switch overhead.",
    },
    "Priority": {
        "name": "Priority Scheduling",
        "type": "Non-Preemptive",
        "desc": "CPU is allocated to the highest-priority process (lowest number). Can suffer from starvation of low-priority processes.",
    },
    "RR": {
        "name": "Round Robin",
        "type": "Preemptive",
        "desc": "Each process gets a fixed time quantum. After expiry, the process is preempted and re-queued. Provides fair CPU sharing and good response time.",
    },
    "MLFQ": {
        "name": "Multi-Level Feedback Queue",
        "type": "Preemptive",
        "desc": "Multiple queues (Q0→Q1→Q2) with decreasing priority and increasing quantum. Processes that use their full quantum are demoted.",
    },
}

PRESETS = {
    "Default": [
        ProcessInput("P1", 0, 8, 3), ProcessInput("P2", 1, 4, 1),
        ProcessInput("P3", 2, 9, 4), ProcessInput("P4", 3, 5, 2),
    ],
    "CPU Bound": [
        ProcessInput("P1", 0, 15, 2), ProcessInput("P2", 2, 12, 3),
        ProcessInput("P3", 4, 18, 1), ProcessInput("P4", 6, 10, 4),
        ProcessInput("P5", 8, 14, 2),
    ],
    "I/O Bound": [
        ProcessInput("P1", 0, 2, 1), ProcessInput("P2", 1, 3, 2),
        ProcessInput("P3", 2, 1, 3), ProcessInput("P4", 3, 4, 1),
        ProcessInput("P5", 4, 2, 2), ProcessInput("P6", 5, 3, 3),
    ],
    "Equal Burst": [
        ProcessInput("P1", 0, 5, 1), ProcessInput("P2", 0, 5, 2),
        ProcessInput("P3", 0, 5, 3), ProcessInput("P4", 0, 5, 4),
    ],
    "Heavy Load": [
        ProcessInput("P1", 0, 10, 3), ProcessInput("P2", 0, 5, 1),
        ProcessInput("P3", 1, 8, 4), ProcessInput("P4", 1, 3, 2),
        ProcessInput("P5", 2, 7, 5), ProcessInput("P6", 3, 6, 1),
        ProcessInput("P7", 4, 4, 3), ProcessInput("P8", 5, 9, 2),
    ],
}


def _compute_metrics(processes: List[ProcessInput], execution_log: List[ExecBlock]) -> SimulationResult:
    completion_times: Dict[str, int] = {}
    first_exec_times: Dict[str, int] = {}

    for block in execution_log:
        if block.process_id != "IDLE":
            completion_times[block.process_id] = max(
                completion_times.get(block.process_id, 0), block.end_time
            )
            if block.process_id not in first_exec_times:
                first_exec_times[block.process_id] = block.start_time

    metrics = []
    for p in processes:
        ct = completion_times.get(p.id, 0)
        tat = ct - p.arrival_time
        wt = tat - p.burst_time
        rt = first_exec_times.get(p.id, p.arrival_time) - p.arrival_time
        metrics.append(ProcessMetrics(
            id=p.id, completion_time=ct,
            turnaround_time=max(0, tat), waiting_time=max(0, wt),
            response_time=max(0, rt),
        ))

    n = len(metrics) if metrics else 1
    avg_tat = sum(m.turnaround_time for m in metrics) / n
    avg_wt = sum(m.waiting_time for m in metrics) / n
    avg_rt = sum(m.response_time for m in metrics) / n

    # Compress consecutive blocks
    compressed: List[ExecBlock] = []
    for block in execution_log:
        if (compressed and compressed[-1].process_id == block.process_id
                and compressed[-1].end_time == block.start_time):
            compressed[-1].end_time = block.end_time
        else:
            if block.end_time > block.start_time:
                compressed.append(ExecBlock(block.process_id, block.start_time, block.end_time))

    # CPU Utilization
    total_time = compressed[-1].end_time if compressed else 0
    idle_time = sum(b.end_time - b.start_time for b in compressed if b.process_id == "IDLE")
    cpu_util = ((total_time - idle_time) / total_time * 100) if total_time > 0 else 0

    # Throughput
    throughput = len(processes) / total_time if total_time > 0 else 0

    # Context switches
    ctx = 0
    last_proc = None
    for b in compressed:
        if b.process_id != "IDLE":
            if last_proc is not None and last_proc != b.process_id:
                ctx += 1
            last_proc = b.process_id

    return SimulationResult(compressed, metrics, avg_tat, avg_wt, avg_rt, cpu_util, throughput, ctx)


def simulate_fcfs(processes):
    current_time = 0
    log = []
    sorted_p = sorted(processes, key=lambda p: (p.arrival_time, int(''.join(filter(str.isdigit, p.id)) or '0')))
    for p in sorted_p:
        if current_time < p.arrival_time:
            log.append(ExecBlock("IDLE", current_time, p.arrival_time))
            current_time = p.arrival_time
        log.append(ExecBlock(p.id, current_time, current_time + p.burst_time))
        current_time += p.burst_time
    return _compute_metrics(processes, log)


def simulate_sjf(processes):
    current_time = 0
    log = []
    remaining = deepcopy(processes)
    while remaining:
        arrived = [p for p in remaining if p.arrival_time <= current_time]
        if not arrived:
            na = min(p.arrival_time for p in remaining)
            log.append(ExecBlock("IDLE", current_time, na))
            current_time = na
            continue
        arrived.sort(key=lambda p: (p.burst_time, p.arrival_time))
        p = arrived[0]
        log.append(ExecBlock(p.id, current_time, current_time + p.burst_time))
        current_time += p.burst_time
        remaining = [r for r in remaining if r.id != p.id]
    return _compute_metrics(processes, log)


def simulate_priority(processes):
    current_time = 0
    log = []
    remaining = deepcopy(processes)
    while remaining:
        arrived = [p for p in remaining if p.arrival_time <= current_time]
        if not arrived:
            na = min(p.arrival_time for p in remaining)
            log.append(ExecBlock("IDLE", current_time, na))
            current_time = na
            continue
        arrived.sort(key=lambda p: (p.priority, p.arrival_time))
        p = arrived[0]
        log.append(ExecBlock(p.id, current_time, current_time + p.burst_time))
        current_time += p.burst_time
        remaining = [r for r in remaining if r.id != p.id]
    return _compute_metrics(processes, log)


def simulate_srtf(processes):
    current_time = 0
    log = []
    remaining = [{"id": p.id, "arrival_time": p.arrival_time, "burst_time": p.burst_time,
                  "priority": p.priority, "remaining_time": p.burst_time} for p in processes]
    while remaining:
        arrived = [p for p in remaining if p["arrival_time"] <= current_time]
        if not arrived:
            na = min(p["arrival_time"] for p in remaining)
            log.append(ExecBlock("IDLE", current_time, na))
            current_time = na
            continue
        arrived.sort(key=lambda p: (p["remaining_time"], p["arrival_time"]))
        p = arrived[0]
        log.append(ExecBlock(p["id"], current_time, current_time + 1))
        p["remaining_time"] -= 1
        current_time += 1
        if p["remaining_time"] == 0:
            remaining = [r for r in remaining if r["id"] != p["id"]]
    return _compute_metrics(processes, log)


def simulate_rr(processes, time_quantum):
    current_time = 0
    log = []
    remaining = [{"id": p.id, "arrival_time": p.arrival_time, "burst_time": p.burst_time,
                  "priority": p.priority, "remaining_time": p.burst_time} for p in processes]
    remaining.sort(key=lambda p: p["arrival_time"])
    queue = []
    not_arrived = list(remaining)

    def enqueue_arrived(time):
        nonlocal not_arrived
        arrived = [p for p in not_arrived if p["arrival_time"] <= time]
        arrived.sort(key=lambda p: p["arrival_time"])
        for p in arrived:
            queue.append(p)
            not_arrived = [rp for rp in not_arrived if rp["id"] != p["id"]]

    enqueue_arrived(current_time)
    while queue or not_arrived:
        if not queue:
            na = not_arrived[0]["arrival_time"]
            log.append(ExecBlock("IDLE", current_time, na))
            current_time = na
            enqueue_arrived(current_time)
            continue
        p = queue.pop(0)
        exec_time = min(p["remaining_time"], time_quantum)
        log.append(ExecBlock(p["id"], current_time, current_time + exec_time))
        current_time += exec_time
        p["remaining_time"] -= exec_time
        enqueue_arrived(current_time)
        if p["remaining_time"] > 0:
            queue.append(p)
    return _compute_metrics(processes, log)


def simulate_mlfq(processes, q0_quantum=2, q1_quantum=4):
    current_time = 0
    log = []
    remaining = [{"id": p.id, "arrival_time": p.arrival_time, "burst_time": p.burst_time,
                  "priority": p.priority, "remaining_time": p.burst_time} for p in processes]
    remaining.sort(key=lambda p: p["arrival_time"])
    q0, q1, q2 = [], [], []
    not_arrived = list(remaining)

    def enqueue_arrived(time):
        nonlocal not_arrived
        arrived = [p for p in not_arrived if p["arrival_time"] <= time]
        arrived.sort(key=lambda p: p["arrival_time"])
        for p in arrived:
            q0.append(p)
            not_arrived = [rp for rp in not_arrived if rp["id"] != p["id"]]

    enqueue_arrived(current_time)
    while q0 or q1 or q2 or not_arrived:
        if not q0 and not q1 and not q2:
            na = min(p["arrival_time"] for p in not_arrived)
            log.append(ExecBlock("IDLE", current_time, na))
            current_time = na
            enqueue_arrived(current_time)
            continue
        if q0:
            p = q0.pop(0)
            exec_time = min(p["remaining_time"], q0_quantum)
            log.append(ExecBlock(p["id"], current_time, current_time + exec_time))
            current_time += exec_time
            p["remaining_time"] -= exec_time
            enqueue_arrived(current_time)
            if p["remaining_time"] > 0:
                q1.append(p)
        elif q1:
            p = q1.pop(0)
            t = 0
            while t < q1_quantum and p["remaining_time"] > 0:
                log.append(ExecBlock(p["id"], current_time, current_time + 1))
                current_time += 1
                p["remaining_time"] -= 1
                t += 1
                enqueue_arrived(current_time)
                if q0:
                    break
            if p["remaining_time"] > 0:
                if t < q1_quantum:
                    q1.insert(0, p)
                else:
                    q2.append(p)
        elif q2:
            p = q2.pop(0)
            log.append(ExecBlock(p["id"], current_time, current_time + 1))
            current_time += 1
            p["remaining_time"] -= 1
            enqueue_arrived(current_time)
            if p["remaining_time"] > 0:
                q2.insert(0, p)
    return _compute_metrics(processes, log)


def run_simulation(processes, algo, rr_quantum=2, q0_quantum=2, q1_quantum=4):
    if not processes:
        return SimulationResult([], [], 0.0, 0.0, 0.0, 0.0, 0.0, 0)
    if algo == "FCFS":
        return simulate_fcfs(processes)
    elif algo == "SJF":
        return simulate_sjf(processes)
    elif algo == "SRTF":
        return simulate_srtf(processes)
    elif algo == "Priority":
        return simulate_priority(processes)
    elif algo == "RR":
        return simulate_rr(processes, rr_quantum)
    elif algo == "MLFQ":
        return simulate_mlfq(processes, q0_quantum, q1_quantum)
    return simulate_fcfs(processes)


def get_state_at_tick(tick, processes, execution_log):
    """Get running/ready/completed/pending process state at a given tick."""
    current_block = None
    for b in execution_log:
        if b.start_time <= tick < b.end_time:
            current_block = b
            break
    running = current_block.process_id if current_block and current_block.process_id != "IDLE" else None

    executed_time = {}
    for b in execution_log:
        if b.process_id == "IDLE":
            continue
        eff_end = min(b.end_time, tick)
        if eff_end > b.start_time:
            executed_time[b.process_id] = executed_time.get(b.process_id, 0) + (eff_end - b.start_time)

    completed, ready_queue, not_arrived = [], [], []
    for p in processes:
        ex = executed_time.get(p.id, 0)
        if ex >= p.burst_time:
            completed.append(p.id)
        elif p.arrival_time > tick:
            not_arrived.append(p.id)
        elif p.id != running:
            ready_queue.append(p.id)
    return {"running": running, "ready": ready_queue, "completed": completed, "pending": not_arrived}
